improver/memprofile.py

```diff
@@ -0,0 +1,129 @@
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing maximum memory profiling utilities."""
+
+import tracemalloc
+from queue import Queue
+from threading import Thread
+from resource import getrusage, RUSAGE_SELF
+from datetime import datetime
+import sys
+
+
+def memory_profile_start(outfile_prefix):
+    """Starts the memory tracking profiler.
+
+    Args:
+        outfile_prefix (str):
+            Prefix for the generated output. 2 files will
+            be generated: *_SNAPSHOT and *_MAX_TRACKER.
+
+    Returns:
+        Active Thread tracking the memory.
+        Active Queue for communication to the thread.
+    """
+    queue = Queue()
+    thread = Thread(target=memory_monitor, args=(queue,
+                                                 outfile_prefix))
+    thread.start()
+    return thread, queue
+
+
+def memory_profile_end(queue, thread):
+    """Ends the memory tracking profiler.
+
+    Args:
+        queue (queue.Queue):
+            Active queue instance to communicate with active thread.
+        thread (thread.Thread):
+            Active thread instance running memory tracking.
+    """
+    queue.put('Stop')
+    thread.join()
+
+
+def memory_monitor(queue, outfile_prefix):
+    """Function to track memory usage, should be run in a separate
+    thread to the main program.
+
+    Samples max_rss, if the max_rss is higher than the previous
+    max_rss, then creates a tracemalloc snapshot. There is a
+    performance overhead when using this.
+
+    Args:
+        queue (queue.Queue):
+            Active queue instance to communicate with the thread.
+        outfile_prefix (str):
+            Prefix for the generated output. 2 files will
+            be generated: *_SNAPSHOT and *_MAX_TRACKER.
+    """
+    tracemalloc.start()
+    old_max = 0
+    snapshot = None
+
+    fout = open("{}_MAX_TRACKER".format(outfile_prefix), 'w')
+    b2mb = 1 / 1048576
+    if sys.platform == 'linux':
+        # linux outputs max_rss in KB not B
+        b2mb = 1 / 1024
+
+    while True:
+        if queue.empty():
```

Comment:
There's no longer a timeout or delay on this queue check, so the effect will be for this thread to spin in a tight loop checking `max_rss` against `old_max`.
This has significant effects on performance. This is likely due to the profiling thread causing python GIL contention with the main thread doing the calculation being profiled.

Some quick timings of an `improver threshold` command on some visibility data from MOGREPS-UK:
- no memory profiling:  2.3 seconds
- memory profiling from earlier commit in this branch with 0.1 second sampling: 5.5 seconds
- current memory profiling method: 62.9 seconds

I think the ~2x slower performance in the earlier commits was manageable, but ~30x slower will make it annoying to use this memory profiling.
Suggest going back to a sampling approach to in order to reduce the performance impact.

Reply:
Okay, I'll re-implement a "wait time"